### DISCUSSION TITLE
问几个问题，关于gae有效性、软件信任自签证书

### DISCUSSION THREAD

**Discussion by cindymao56** [SeaHOH/GotoX#153] [问答/咨询] [answered]
作者您好，我有几个问题

## 1

Google最近给我们发了邮件
> [Reminder] Google App Engine Standard apps still running on Go versions 1.9 and earlier will be shut down June 30, 2020

>  Migrate your App Engine Standard apps to Go version 1.11 or
> above by June 30, 2020.
> Hello Google App Engine Customer,
> 
> On October 1, 2019, we discontinued support for Google App Engine Go versions 1.9 and earlier. Applications that were already deployed on these versions continued to run, but no new deployments were allowed. On June 30, 2020, applications still running on unsupported versions of Go will be shut down.
> What do I need to do?
> 
> To continue running your App Engine Standard applications, redeploy these apps using Go 1.11 or above before June 30, 2020:
> 
>     If your applications use the App Engine Go SDK APIs, upgrade to Go 1.11.
>     If you are not using these APIs, upgrade to Go 1.12.

1. 您仓库中gae服务端那一分支是两年前更新的，那么这个工具配合gae方案目前是不可用的吗？
1. gae服务端代码有没有更新的打算？

## 2

我看到readme中
> 由于平台限制，对于通过 GAE 的 https 流量，GotoX 使用自动生成的证书作为凭证，采取中间人方法进行代理；对于需要修改（某些自动代理规则需要）的 https 流量也是如此，不论其是否通过 GAE。
为防止被滥用，谷歌在 GAE 代理的头字段中会包含了你使用的 AppID 和原始请求的 TraceID 信息，请慎记之。
以上描述也适用于 CFWorkers。

对此我有几个疑问：
1. 对于直接连接的google服务，gotox为什么也要求用户信任自签证书？
1. Cloudflare Worker也在代理字段中带类似appid的信息吗？
1. Cloudflare Worker作代理应该仅需要用户信任Cloudflare的证书（按照我理解cloudflare文档，不一定对），似乎gotox也要求信任自签证书

谢谢

**Comment by SeaHOH** [comment, marked as answer]
> * 您仓库中gae服务端那一分支是两年前更新的，那么这个工具配合gae方案目前是不可用的吗？
> 
> * gae服务端代码有没有更新的打算？

我已多次在不同位置声明过不再维护，现有服务端也无法直接使用，本地客户端相关代码仍然保持正常提供功能，如果服务端没人跟进，一段时间后就会移除它们。

>     1. 对于直接连接的google服务，gotox为什么也要求用户信任自签证书？

这个没有限死，可以修改自动规则使用转发连接。这么做，部分被屏蔽的 SNI 会导致你的网络被自动限制 (影响不定视运气)，除非只使用 ESNI，但是这样还需要走 GotoX？直接直连不好吗。然后，不同域名连接也无法相互复用，需要非常多的 IP 才能正常连接，然而并没有那么多 IP 给你使用。而且因为丢包率高，容易出现卡连接的现象。

>     2. Cloudflare Worker也在代理字段中带类似appid的信息吗？

是的，你可以看它的文档，或者抓包查看。

>     3. Cloudflare Worker作代理应该仅需要用户信任Cloudflare的证书（按照我理解cloudflare文档，不一定对），似乎gotox也要求信任自签证书

如果你的需求仅只是浏览，那么 CFWorkers 使用几行简单代码就能反代，结果就是除了浏览你什么都没法做，而且还只能一对一反代，稍微复杂些的可以一对多。你可以看看其它使用 CFWorkers 的非网页代理 (暂时我只看到一个)，同样需要有自签证书。

关于自签证书，采用的是 2048 位密钥，近期内无需担心安全性。前提是本地密钥要保管好，我在证书部分有说明。而且你也不会用 GotoX 来犯罪，对吧。

**Comment by Quantum2Bits** [comment]
GAE现在还可以用，好好的。以前部署的不用改，可以继续用。服务器也可以现在部署，但要修改一点代码，就是代码开头的一点点，很好改。但是，现在部署需要绑定支付方案。

**Comment by SeaHOH** [comment]
> 现在部署需要绑定支付方案。

对，我就是因为这个原因不愿意继续使用。以前部署的近期都会 down 掉。

欢迎 PR，如果有人愿意维护 GAE 服务端，我就不会在以后的版本中移除本地支持。

**Comment by Quantum2Bits** [comment]
以前部署的服务器不能用了。重新部署后还是不行，我用的是老的appengine package。如果用新的appengine package编译通不过。

**Comment by SeaHOH** [comment]
旧版 appengine 支持会被移除，新版 appengine 移除了一些 fetch 相关的对象，现在直接使用标准库就可以，迁移文档有详细说明。

**Comment by Quantum2Bits** [comment]
用新的appengine package 做一些修改后可以部署。可以运行，比如可以用浏览器看app engine 的网址，但是与客户端不能合作，估计是服务器的一些行为变了。

不用用新的appengine package也可以部署，可以运行。这样的好处是可以在任何主机上运行，包括本地测试。只是和客户端不能合作。应为不了解客户端的工作细节。

**Comment by SeaHOH** [comment]
你是要继续使用 GCP 吗？你 PR 的话我可以帮忙 code review，然后 test&debug 还是需要你自己来。

**Comment by Quantum2Bits** [comment]
> 你是要继续使用 GCP 吗？你 PR 的话我可以帮忙 code review，然后 test&debug 还是需要你自己来。

[gae.go.txt](https://github.com/SeaHOH/GotoX/files/5107110/gae.go.txt)

这是用新的appengine package的代码。他的监听端口由port环境变量获取，不能自己定义。几乎没有什么改动，只是把appengine.Context改成了context.Context。

**Comment by Quantum2Bits** [comment]
> 你是要继续使用 GCP 吗？你 PR 的话我可以帮忙 code review，然后 test&debug 还是需要你自己来。

[gaej.go.txt](https://github.com/SeaHOH/GotoX/files/5107209/gaej.go.txt)
这是不用appengine package的代码。把appengine.Context改成了context.Context。urlfetch.Transport 改成了http.Transport。可以在本地运行，也可以上传到GAE。这个版本不好获取appengine的一些参数。其他的都一样可以跑。

**Comment by SeaHOH** [comment]
迁移参照 https://cloud.google.com/appengine/docs/standard/go/go-differences ，你最好 PR，这里不方便继续讨论。

**Comment by Quantum2Bits** [comment]
PR是指什么？我用email发给你了，不知能收到否？

**Comment by SeaHOH** [comment]
> PR是指什么？

Pull request (拉取请求)，即申请拉取某个 repo (仓库) 的不同版本，经过 code review (代码检查) 后决定是否 merge (合并) 此 pull 到目标 repo。Github PR 有提供 code review 功能，使用它更加方便有条理，否则很多时候会浪费不必要的精力来考虑如何引用以及理解混乱的交流。
1. 点击页面右上角 fork GotoX 到自己的 repo。
1. 点击左上角从 master 分支切换到 gaeserver.goporxy 分支。
1. 修改保存或上传文件。
1. 点击页面上方新出现的 Pull request 按钮。

如果还看不明白，请搜索网上教程。

**Comment by Quantum2Bits** [comment]
搞好了，在main()里面加上一句appengine.Main()就能用了。估计appengine package过一些时间又要升级。所以还是需要一个不用appengine packag版本才好。

**Comment by SeaHOH** [comment]
参照上面的链接，不要使用 appengine。
还有，如果你不愿意 PR，继续在这里讨论也没什么意义，迁移指南写得很详细，你都没有照着修改。

**Comment by Quantum2Bits** [comment]
> 参照上面的链接，不要使用 appengine。
> 还有，如果你不愿意 PR，继续在这里讨论也没什么意义，迁移指南写得很详细，你都没有照着修改。

已经PR了，gae.go 能用，gaej.go 有问题，这里只有一句appengine.Main()是用appengine的。否则客户端连不上。

**Comment by SeaHOH** [comment]
https://github.com/Quantum2Bits/GotoX/pull/1
加个链接，愿意帮忙的可以参与。

**Comment by SeaHOH** [comment]
GAE 相关问题我不会再直接参与改进，你可以请其他人帮忙。
